### ISSUE TITLE
patch for finger14.py

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10365]
|<img alt="mwaite's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| mwaite reported|
|-|-|
|Trac ID|trac#436|
|Type|defect|
|Created|2003-12-16 22:07:06Z|

Attachments:

* [finger.tac](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6c4/6c4c04be8f82a4e053bde03dd716d59c841cfda9/deaa4eaa9971a166aed31dde0d93be636a0ac5d6.tac) (1568 bytes) - added by mint on 2004-03-21 10:32:39Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__436 436
type__defect defect
reporter__mwaite mwaite
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ 
keywords__documentation documentation
time__1071612426000000 1071612426000000
changetime__1472068444945208 1472068444945208
version__ 
owner__hypatia hypatia
cc__moshez cc__hypatia cc__mwaite cc__mint
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="mwaite's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|mwaite commented|
|-|-|

```
#!html
<pre>
Hi,

Below is a patched up version of finger14.py for the 
tutorial page.

The key difference here is that the FingerService 
is registered as a service of the application, and so 
the startService() and stopService() methods are called 
appropriately.

I have also removed the statement 
   f.startService = self.startService
from within getFingerFactory(), because it didn't 
do anything.  Also, self.file has become self.filename
in _read().

It does work now, but is this the right way to do it ?

cheers,
Martin

=======================================

# Read from file
from twisted.application import internet, service
from twisted.internet import protocol, reactor, defer
from twisted.protocols import basic

class FingerProtocol(basic.LineReceiver):
    def lineReceived(self, user):
        self.factory.getUser(user
        ).addErrback(lambda _: "Internal error in server"
        ).addCallback(lambda m:
                      (self.transport.write(m+"\r\n"),
                       self.transport.loseConnection()))

class FingerService(service.Service):
    def __init__(self, filename):
        self.users = {}
        self.filename = filename
    def startService(self):
        def _read(self):
            for line in file(self.filename):
                user, status = line.split(':', 1)
                user=user.strip()
                status=status.strip()
                self.users[user] = status
            self.call = reactor.callLater(30, _read)
        _read(self)
        service.Service.startService(self)
    def stopService(self):
        service.Service.stopService(self)
        self.call.cancel()
    def getUser(self, user):
        return defer.succeed(self.users.get(user, "No such user"))
    def getFingerFactory(self):
        f = protocol.ServerFactory()
        f.protocol, f.getUser = FingerProtocol, self.getUser,
        return f

application = service.Application('finger', uid=1, gid=1)
f = FingerService('/etc/users')
finger = internet.TCPServer(79, f.getFingerFactory())

f.setServiceParent(service.IServiceCollection(application)) 
finger.setServiceParent(service.IServiceCollection(application))
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement commented|
|-|-|

```
#!html
<pre>
Moshe, can you check the patch? I can do it, but you may be faster.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement commented|
|-|-|

```
#!html
<pre>
Applied relevant parts of the modified code (for example, didn't apply the
inline _read()). Thanks for the patch -- however it would be better to attach
future patches as files rather than putting them in the comments, and even
better to attach a unified diff (diff -u) against CVS.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="mint's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|mint commented|
|-|-|

```
#!html
<pre>
I think this file is still broken, at least on the website:
http://www.twistedmatrix.com/documents/current/howto/listings/finger/finger14.py

FingerService is still not registered as a service of the application (~ line 43).
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
The website docs are only updated when we make a release (otherwise the docs
would describe features that don't exist in a release!).  You can view the
latest docs directly in subversion at:
http://cvs.twistedmatrix.com/cvs/trunk/doc/howto/listings/finger/finger14.py?view=auto&amp;rev=9982&amp;root=Twisted

You can also see the latest build of the docs at
http://twistedmatrix.com/users/warner/doc-latest/, generated by the buildbot.

Marking as resolved, again.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

In changeset 7133de6b678c101e3a62119801ce0c7579f0f7d2
```
#!CommitTicketReference repository="" revision="7133de6b678c101e3a62119801ce0c7579f0f7d2"
Merge pull request #436 from twisted/port-pb-7598

Author: hawkowl, glyph

Reviewer: rodrigc

Fixes: #7598 

Port Perspective Broker to Python 3.
```
